Internal: Arrangement of style transformers [EDS-399]

ronros-elementor opened the pull request:
Adding Styles_Renderer to the Post_CSS rendering flow

+ Organized the first two style transformers, though expected to be changed later on once we have a solid implementation for Transformer_Base

ronros-elementor committed:
wip

ronros-elementor committed:
remove redundant comments

ronros-elementor committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
Merge remote-tracking branch 'origin/main' into feature/EDS-349

ronros-elementor committed:
Merge remote-tracking branch 'ronros-elementor/feature/EDS-349' into feature/EDS-349

ronros-elementor committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
added snapshots assertions

elementorbot committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
composer update

ronros-elementor committed:
Merge remote-tracking branch 'ronros-elementor/feature/EDS-349' into feature/EDS-349

ronros-elementor committed:
downgraded php to 7.4

ronros-elementor committed:
lint

elementorbot committed:
Merge branch 'main' into feature/EDS-349

elementorbot committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
CR fixes (+ downgraded php to 7.4 to resolve composer packages issues)

ronros-elementor committed:
Merge remote-tracking branch 'ronros-elementor/feature/EDS-349' into feature/EDS-349

ronros-elementor committed:
should be `unset` if not valid value, to match the TS api

ronros-elementor committed:
wip

ronros-elementor committed:
tests

ronros-elementor committed:
changed to fn

ronros-elementor committed:
removed redundant type declaration

elementorbot committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
wip

ronros-elementor committed:
Merge remote-tracking branch 'ronros-elementor/feature/EDS-349' into feature/EDS-349

ronros-elementor committed:
added test of variants with state and breakpoint

ronros-elementor committed:
and snapshot

ronros-elementor committed:
and snapshot!

ronros-elementor committed:
changed render arguments to be an array

ronros-elementor committed:
Merge branch 'main' into feature/EDS-349

elementorbot committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
moved arguments to be declared inline

elementorbot committed:
Merge branch 'main' into feature/EDS-349

elementorbot committed:
Merge branch 'main' into feature/EDS-349

elementorbot committed:
Merge branch 'main' into feature/EDS-349

elementorbot committed:
Merge branch 'main' into feature/EDS-349

ronros-elementor committed:
wip

ronros-elementor committed:
fix wrong type for breakpoints

ronros-elementor committed:
removed format script from composer

ronros-elementor committed:
merge main/eds-349

ronros-elementor committed:
lint

ronros-elementor committed:
merge main

ronros-elementor committed:
remove format

ronros-elementor committed:
lint

elementorbot committed:
Merge branch 'main' into feature/EDS-399

ronros-elementor committed:
tests

ronros-elementor committed:
Merge remote-tracking branch 'origin/feature/EDS-399' into feature/EDS-399

ronros-elementor committed:
wip

elementorbot committed:
Merge branch 'main' into feature/EDS-399

elementorbot committed:
Merge branch 'main' into feature/EDS-399

elementorbot committed:
Merge branch 'main' into feature/EDS-399

elementorbot committed:
Merge branch 'main' into feature/EDS-399

elementorbot committed:
Merge branch 'main' into feature/EDS-399

elementorbot committed:
Merge branch 'main' into feature/EDS-399

Nevoss commented on modules/atomic-widgets/base/style-transformer-base.php:
```diff
@@ -0,0 +1,24 @@
+	exit; // Exit if accessed directly.
+}
+
+abstract class Style_Transformer_Base {
```
I think this will be handle by my side, but we can talk about it

ronros-elementor replied:
Totally agree, it's a placeholder for now but I based it on your work
One thing I need to align with you on is how can I invoke the value-transform logic from within a transformer (I passed a callback as its second argument)

ronros-elementor replied:
Created [this task](https://elementor.atlassian.net/browse/EDS-424) to refactor once the main backend transformers task is done

Nevoss commented on modules/atomic-widgets/module.php:
```diff
@@ -51,6 +68,12 @@ public function __construct() {
+				'elementor/atomic-widgets/styles/transformers',
+				fn ( array $transformers) => $this->register_style_transformers( $transformers )
+			);
+			add_action( 'elementor/element/parse_css', fn( Post $post, Element_Base $element ) => $this->parse_atomic_widget_css( $post, $element ), 10, 2 );
```
Please move all the logic of the rendering to a different class, something like `Element_Parse_CSS`, the action could be here in the module but the logic can live outside

ronros-elementor replied:
Done

Nevoss commented on modules/atomic-widgets/module.php:
```diff
@@ -110,4 +147,106 @@ private function enqueue_scripts() {
+		return $this->breakpoints;
+	}
+
+	private function build_breakpoints(): array {
```
Why do you need normalize the breakpoints?

ronros-elementor committed:
merge main

ronros-elementor committed:
merge main

ronros-elementor committed:
Merge branch 'feature/EDS-399' of https://github.com/elementor/elementor into feature/EDS-399

ronros-elementor committed:
extracting logic to a designated class

ronros-elementor committed:
remove composer format

ronros-elementor committed:
removed breakpoints normalization

ronros-elementor committed:
Merge remote-tracking branch 'origin/main' into feature/EDS-399

ronros-elementor committed:
fix for tests - added expect calls + throwing exceptions from within the transformers (though irrelevant as they shouldn't be used directly)

+ it will anyway change to a slightly different structure once a proper base transformer class is merged

ronros-elementor committed:
wip

ronros-elementor committed:
wip

ronros-elementor committed:
lint

ronros-elementor committed:
Merge branch 'main' into feature/EDS-399

ronros-elementor committed:
removed redundant tests and try-catch statements

ronros-elementor committed:
Merge remote-tracking branch 'origin/feature/EDS-399' into feature/EDS-399

ronros-elementor committed:
changed mock class declaration in test

Omerisra6 commented on modules/atomic-widgets/module.php:
```diff
@@ -34,6 +34,7 @@ public function __construct() {
 
 		if ( Plugin::$instance->experiments->is_feature_active( self::EXPERIMENT_NAME ) ) {
 			( new Dynamic_Tags() )->register_hooks();
+			( new Widget_Styles() )->register_hooks();
```
I think the atomic widgets module should register the hook for the widgets styles.

The `Widget_Styles` module should be a generic module that takes a style object and returns the css using the transformers.

ronros-elementor replied:
As discussed - since Styles_Renderer _is_ already that generic module which gets a styles array and returns a css string - adding the handler for `elementor/element/parse_css` within the general module might cause some unnecessary "noise" to that module.
But I did separate, within the `Atomic_Styles` module, the logic of the action handler, and the use of `Style_Renderer` into two different functions for better readability

Omerisra6 commented on modules/atomic-widgets/widget-styles.php:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * @var array<string, Style_Transformer_Base> $transformers
+	 */
+	private array $style_transformers = [];
```
I think it makes more sense that the transformers will be stored in the `Styles_Renderer`, I am not sure why we pass the renderer a config instead of storing it there...

ronros-elementor committed:
wip

ronros-elementor committed:
wip

ronros-elementor committed:
Merge branch 'main' into feature/EDS-399

ronros-elementor committed:
tests

ronros-elementor committed:
Merge remote-tracking branch 'origin/feature/EDS-399' into feature/EDS-399

ronros-elementor committed:
tests

ronros-elementor commented on modules/atomic-widgets/atomic-styles.php:
```diff
@@ -0,0 +1,118 @@
+		] );
+	}
+
+	private function parse_element_style( Post $post, Element_Base $element ) {
+		if ( ! ( $element instanceof Atomic_Widget_Base ) || Post::class !== get_class( $post ) ) {
+			return;
+		}
+
+		$styles = $element->get_raw_data()['styles'];
+
+		if ( empty( $styles ) ) {
+			return;
+		}
+
+		$post->get_stylesheet()->add_raw_css( $this->convert_styles_to_css( $styles ) );
```
Done

ronros-elementor committed:
renamed variables

ronros-elementor commented on modules/atomic-widgets/styles/styles-renderer.php:
```diff
@@ -132,15 +134,21 @@ private function wrap_with_media_query( string $breakpoint, string $css ): strin
 			return $css;
 		}
 
+		$breakpoint_config = $this->breakpoints[ $breakpoint ];
```
Good point, but lets settle on `breakpoint_id` instead of `breakpoint_name`, as this is its identifier 🙂 (and of course the `config` is dropped)

ronros-elementor committed:
adding test to check widget is atomic widget

Omerisra6 committed:
Merge branch 'main' into feature/EDS-399
